Import session

phweyland opened the pull request:
Attempt to partially fix #2762, partially fix #8292, fix #6363 and fix #6493.

![image](https://user-images.githubusercontent.com/23012047/109499103-355c4800-7a73-11eb-96ab-d94c17e40d19.png)

![image](https://user-images.githubusercontent.com/23012047/109506291-3b0a5b80-7a7c-11eb-81a2-6164fd378a46.png)


Replace the previous `image...` button behavior.
It is able to import files where they are as today or use the session pattern to import from other folders.
The main difference I can see is to not open darkroom anymore at the end of the import (everything is done in a separated job).

It covers also all the features of `folder...` button. However I haven't removed it because the ui is a bit different and some people could be sensitive to that. The new ui displays all imported files. In case of recursive `folder...` import, this can represent thousands of files. On the other hand the `folder...` button just shows the root folder.
 
Changes:
- displays thumbs on demand
- shows file date/time
- can import & copy files using session pattern as camera import does
- reads usb memory card (but not the cameras (1))
- doesn't use libgphoto2 directly

(1) all file browsers read the cameras as well (thanks to libgphoto2 I think) but  gtkfilechooser seems not to be able to do the same.

Comments, suggestions are welcome.

TurboGit reviewed (changes requested):
I have many dt crashes with this PR. I tried to look for a backtrace but nothing obvious as I have only Gtk+ traces, could maybe be a signal break or some memory corruption. It happens when I import, select and remove image. Sometimes only importing crash.

EDIT: using the import [folder...] only as available on master.

phweyland commented:
> I have many dt crashes with this PR

Aie ...

> using the import [folder...] only as available on master.

Normally I haven't touched anything on [folder...] side, only [image...]. I don't understand this, sorry.

TurboGit commented:
I just do import a folder as it is today with `[folder...]` and I have lot of crashes indeed. Maybe some side effects? Can't you reproduce? I did that with some folder not already imported or folder already imported. In both cases dt is very very unstable....

TurboGit commented:
If you can't reproduce I'll see if I can get a better trace or idea about this issue.

phweyland commented:
Thanks. Building again...

phweyland commented:
Yes, with [folder...] I get a crash as well. A bit strange because I haven't changed this part. Looking at it.

Nilvus commented:
Just reading that PR, I love it. You made my day @phweyland. I was waiting for date and hour since a long time as you can see in my linked issue. Many many thanks for that work. Will test ASAP (not sure I will have time before this is merged if merged quickly but no big deal. Just THANK YOU for adding all those features.

phweyland commented:
Fixed the issue with [folder...]. 
Have you experienced crashes with [image...] as well ?

TurboGit commented:
I didn't test `[Image..]`, testing now...

TurboGit commented:
Just tested again, no more crash on the `[Folder...]`. And `[Image...]` is working too.

TurboGit commented on data/darktableconfig.xml.in:
```diff
@@ -734,6 +734,20 @@
+    <name>ui_last/in_situ</name>
+    <type>bool</type>
+    <default>TRUE</default>
+    <shortdescription>import images as they are</shortdescription>
```
import images in-place

TurboGit commented:
I would really be tempted by removing the `[folder..]` button at this stage. It *only* save two clicks. We'll see.

But for sure now the button `[image...]` should have an `s` in image**s**.

johnny-bit commented:
Huh, I need to test this...

AlicVB commented:
Great, I like that change !
Some minor comments : 
1. I find the checkboxes at the line start confusing. Initially, I've thought it was to select the images to import... I'm far from an UI specialist, but for me, in this sort of disposition the initial checkox is usually used to indicate on which elements we want to make an action... Maybe we could replace this with a "load" button in the thumb area ? (not sure it'll be pretty)
2. for my use, I would have selected all the images by default. That would have saved one more click... Isn't the usual workflow "import whole folder" when you select a folder ? (real question)
3. as @TurboGit I think we should remove the "folders..." button, even more with the above change and rename "image..." to "images..." or even "import images..."

phweyland commented:
>     1. I find the checkboxes at the line start confusing.

At least I can move this column just after the thumbnail one. and add a tooltip.


>     2\. I would have selected all the images by default. That would have saved one more click...

yeh.... that's a valid option too...

phweyland commented:
Import again with session data duplicates the images. I would overwrite the images instead.
What do you think ?

AlicVB commented:
> Import again with session data duplicates the images. I would overwrite the images instead.
> What do you think ?

yes, I think this should be the default behavior. I'm not sure to see workflow where we want to duplicate the images, but in this case, we can always add that feature with a new option afterwards...

> At least I can move this column just after the thumbnail one. and add a tooltip.

I don't think this would works, as we would have a blank space at first and the a checkbox... quite strange... I've just done a small mockup of what I have in mind, just as an example : 
![Capture d’écran_2021-03-02_22-21-22](https://user-images.githubusercontent.com/1818223/109718877-21fec900-7ba8-11eb-871f-3fd67ee0759a.png)
- click on the eye to show the thumbnail, click on the thumbnail to hide it, click on the header "aperçu" to show/hide all thumbnails.
- I've also added a reminder of the number of images selected just before the confirmation button. Not sure the window title is enough (not to speak about those who don't have it)

phweyland commented:
> I've just done a small mockup of what I have in mind, just as an example

Like it. I'll what I can do. 

>     * I've also added a reminder of the number of images selected just before the confirmation button

Not sure if I can do this with gtkdialog. I'll look at it.

AlicVB commented:
> Not sure if I can do this with gtkdialog. I'll look at it.

oh yes, that may be difficult... what about including the numbers directly in the "import" button on the fly ? like "import 1/1088 images" That would be even more visible (but at the cost of a "large" button)

phweyland commented:
Here some progress on thumbnails. Remains the repetition of the numbers.

![image](https://user-images.githubusercontent.com/23012047/109737111-40fb5c00-7ba4-11eb-9c04-0505c385246a.png)

phweyland commented:
Moved images counts to action box.
All images selected by default.
I've tried overwrite existing images, but _image_import_internal() doesn't read the file if the image already exists. As the user can (should) clean up the folder before importing again, I've given up on this point (I don't want to create problems elsewhere as well).

![image](https://user-images.githubusercontent.com/23012047/109813502-34125300-7c0c-11eb-9ca7-40dd81cf6c8f.png)

Nilvus commented:
> Just reading that PR, I love it. You made my day @phweyland. I was waiting for date and hour since a long time as you can see in my linked issue. Many many thanks for that work. Will test ASAP (not sure I will have time before this is merged if merged quickly but no big deal. Just THANK YOU for adding all those features.

@pĥweyland: I'm quite disappointed here. I was thinking, as you link my #2762 issue in your first comment that you implemented what I want in import from camera button and by testing your PR, I just discover that... NOT. This PR is awesome but that finally didn't made my day and you should remove "fix #2762" as this PR doesn't fix it. I would like to see exact same thing for import from camera/card (mass media storage) button. Could you add that in that PR or in another?

Nilvus commented:
2 things after testing it a little more:

- We need to think about import from camera button. I always use it to import from SD card but indeed if it's now possible to import form it AND import it on a wanted folder with renaming/creating folder, maybe this button became a sort of duplicate. But of course, it's only for SD card and it remains camera USB connection. If now I import from SD card using this PR, I have that I want but not if I import from camera so. So to be precise: that fix partially #2762.
- second thing: when clicking on import button from this new window, the window keep open while importing is done in background. It's just bad behavior. Most of the times, if I have selected a folder and click on import, I would see the window close automatically after clicking on import. When I had click on import button (after selecting my images in that new windows, for few seconds, I have thought that import doesn't work. I just remember progressbar and see it on bottom left of main window to understand all is good. Not intuitive. But if the window was closed, that would make the user thinking all is good.

phweyland commented:
>     * that fix partially #2762.

That's perfectly correct.

> It's just bad behavior. 

That's a point of view. I feel the opposite. Just because I can import by pieces (by day for example) without opening again and again the dialog. When I'm done, escape key nicely closes the dialog window.

quovadit commented:
I love this feature, thanks very much :tada:

Some comments:
- for me the new feature even replaces the existing 'import from camera'
  - if the camera is mounted as storage device it works perfect
  - you can sort by date/time, and its much faster because only selected thumbnails are loaded
- the button should be called 'import images...'
![01_import-module](https://user-images.githubusercontent.com/5677563/109877303-25ab5200-7c73-11eb-8c20-0605b58077e5.png)
- I would prefer to have the option 'recursive directory' (and possibly ignore jpg) on the dialog where you choose the folder
![02_choose-folder](https://user-images.githubusercontent.com/5677563/109877331-2d6af680-7c73-11eb-9ea4-db7bf3b26562.png)
- a radio button "import images in-place" vs. "copy images" would be much clearer than a checkbox
- I would put the list of files on top and have the parameters underneath
- the option 'override today's date' is missing
![03_file-list](https://user-images.githubusercontent.com/5677563/109877906-df0a2780-7c73-11eb-85d1-a0c5a55f3055.png)
- the 'parameters' could be completely moved to the 'settings' tab
![04_settings](https://user-images.githubusercontent.com/5677563/109877937-e92c2600-7c73-11eb-98d8-91649c593e94.png)

Nilvus commented:
> > It's just bad behavior.
> 
> That's a point of view. I feel the opposite. Just because I can import by pieces (by day for example) without opening again and again the dialog. When I'm done, escape key nicely closes the dialog window.

Of course, but how you deal with the fact that nothing really help the user to know that import have been applied correctly? Thinking of progressbar in bottom left of main window while the dialog remains open is bad UI (and it's not a point of view. We need to think of something else, like an option or dialog to confirm import have been applied and a choice to close or let open for a new set of images to import.

phweyland commented:
>     * if the camera is mounted as storage device it works perfect

This is a good piece of news. But it may not work on other OS, windows ?
I'm exploring the other way, forcing gtkfilechooser to recognize no local folders:
![image](https://user-images.githubusercontent.com/23012047/109881496-2a620d00-7c57-11eb-9e35-2230c23c2c57.png)
but of course we'll need to use libgphoto2 to go further (even the thumb don't work at this point)...

> the button should be called 'import images...'

agreed.

> I would prefer to have the option 'recursive directory' (and possibly ignore jpg) on the dialog where you choose the folder

we avoid to use gtkfilechooser for that purpose (not compatible gtk4). That's why they are where they are.

> a radio button "import images in-place" vs. "copy images" would be much clearer than a checkbox

Possible, but radio button are not really used in dt. An alternative could be a combo box.

> the option 'override today's date' is missing

I did not think it was useful for importing, but now you talk about it, I can at least see a usage when importing scans (without exif date/time)... Do you see other use cases ?

> the 'parameters' could be completely moved to the 'settings' tab

Yes it could. But I don't like to loose the view on these parameters...

phweyland commented:
> Of course, but how you deal with the fact that nothing really help the user to know that import have been applied correctly?

That's certainly not in closing the windows after clicking import.

quovadit commented:
> we avoid to use gtkfilechooser for that purpose (not compatible gtk4).

That's very sad. An alternative would be to have two import buttons:
![01b_import-module_two-buttons](https://user-images.githubusercontent.com/5677563/109918046-ac355300-7cb6-11eb-9ff9-e3f10e4f80b5.png)

Ignore jpg could be shown underneath the file-list:
![03b_file-list_ignore-jpg](https://user-images.githubusercontent.com/5677563/109918543-6dec6380-7cb7-11eb-8580-53334051f511.png)


> even the thumbs don't work at this point

Thumbnails work fine for me even if I import from camera, it's just very slow


> radio button are not really used in dt. An alternative could be a combo box.

Just because there is no radio button so far does not mean we shouldn't use it here.
I think it's important to see both options (in-place / copy) straightaway to understand what this is about.


> 'override today's date' ... Do you see other use cases ?

If one event last longer than one day I want to have the beginning-date for all images so exif-date won't work


> nothing really help the user to know that import have been applied correctly

I would also focus on the main usecase which is using the import once. It's really strange that the window keeps as is after clicking 'import'. And the 'done' button only makes sense after the import.
Here is a possible flow:
![03c_import-again](https://user-images.githubusercontent.com/5677563/109921868-cd993d80-7cbc-11eb-9c9d-c5a1076bea13.png)

AlicVB commented:
About the settings, I think that the main problems is that we have now 2 places to define them : in the import lib panel, and inside the new window (btw, I love what you have done so far).. that's confusing. esp. if we have an unique button.

On the other hand, I agree with you about keeping settings "under the eye". and I would add that the more place we can let to the list the better it is... Maybe a solution can be to use the same "fold/unfold" thing we currently have in the import panel.
- That would have the advantage to allow to set the "recursive" thing afterwards ("just" need to reload the list of images). this is valid for all the settings.
- We can even image a checkox "keep the import window open after the import" because I think both workflow are valid and should be handled

All this start to be a huge (but great) change. Don't hesitate if you need help ;)

phweyland commented:
>     * if the camera is mounted as storage device it works perfect

I don't see any uuid related to the camera. What do you mean by mounted as storage device ? How do you do that ? TIA and sorry for the dumb questions.
If we can mount the camera on the fly in dt we may have a good starting point ...

quovadit commented:
When connecting my camera via usb while darktable is still closed, I can simply mount the camera in my file explorer like any other usb-device:
![mount-camera](https://user-images.githubusercontent.com/5677563/110074171-0995d680-7d81-11eb-9b6b-dbffbfdf9961.png)

In the folder-path you can see, that gphoto2 is used to mount the camera as 'disk':
![mounted](https://user-images.githubusercontent.com/5677563/110074633-d4d64f00-7d81-11eb-9e78-96a22c1c95b7.png)

Some cameras even have an option like "Mass Storage" (also called "USB Mass Storage" or "Mass Storage Device"), then you don't need a separate tool in the background to enable the mounting.

As soon as darktable is started, it tries to mount the camera as 'camera-device' (also used for tethering).

You can only mount the camera as one of the two options at the same time. This is also described in dtdocs:

> import images from a camera
>     To import images from a camera, first connect the camera to your system with a USB cable. If your system tries to automount the camera’s files, you should choose to abort the mount operation, otherwise the camera cannot be accessed from within darktable. If you don’t see your camera listed in the import module, press the “scan for devices” button. Once your camera is detected the import module should offer the ability to import images or tether your camera while shooting.
>

phweyland commented:
@quovadit, thanks for your detailed answer.
I understand better now. 
In my case, the D700 has not the option "mass storage" (I've found users complaining about this). So it behaves a bit differently than your camera (see my [previous snaphost](https://user-images.githubusercontent.com/23012047/109881496-2a620d00-7c57-11eb-9e35-2230c23c2c57.png)).
It appears as well, not as a local folder/device, but as uri. This breaks most of the work that dt makes on files.  That's where libgphoto2 is necessary.

As it's camera dependent, my hope to attach the device differently on the fly has gone...

johnny-bit commented:
Unfortunately gphoto2 also doesn't make import/tethering procedure camera-agnostic so the import/tether from camera stuff via gphoto2 wil be constant pain.

phweyland commented:
Thanks for all the suggestions received. There are always more ideas in several heads than in a single one. :)
I try to digest all this and come back to you.

@Nilvus, would there be a way to have "by default" in CSS a constant height for label, entry and check box (and possibly combo box) ? That would help to get a consistent presentation. In the examples above, every setting line has a different height (with no tweak).

jenshannoschwalm commented:
I just found time to read into this pr .. very very cool!

About libgphoto2, that is really hard stuff and some things are not documented at all or work in one situation and don't in another one. Cameras a direct usb mass storage protocol seem to work fine for all os. Those offering a PTP protocol are somewhat problematic, on linux systems having the gvfs in work (as described above) they are just fine, for windows that's a different story with some open issues that are not understood for us dt folks.

You may look at my latest gphoto related pr's here.

phweyland commented:
> You may look at my latest gphoto related pr's here.

I'll do ! Thanks

phweyland commented:
Some updates:
- move all prefs to import dialog window. removed the module expander 
There are 3 categories of prefs:
-- dialog itself (jpeg, recursive, ...) with list update on the fly
-- mode and patterns (added back the datetime override)
-- metadata
- end of import message (control_log)
- renamed [images...] button
- changed in-place toggle button to modes enumeration (maybe there is a volunteer to add the move mode :)).
- repeat the chosen mode in action-box

Remains
- update presets to wrap all this.
- look at the @jenshannoschwalm gphoto related pr's.

![image](https://user-images.githubusercontent.com/23012047/110172591-201d4b80-7ddc-11eb-9ad7-67c6563b77b2.png)

![image](https://user-images.githubusercontent.com/23012047/110172663-3d521a00-7ddc-11eb-892d-e2b678cbed4a.png)

![image](https://user-images.githubusercontent.com/23012047/110172712-4fcc5380-7ddc-11eb-8216-07d46bc5d11e.png)

quovadit commented:
I love that "recursive" and "ignore jpg" effect the list immediately!
"Keep this window open" is also great, but I would put that checkbox near the import button.
It's absolutely great to have all prefs in the import dialog window now.

The only thing I still don't like is the choice between "in-place" and "copy".
Not only do I have to switch tab, it's also still a drop down.
Adding images in-place or copying them from camera/sd-card are 2 important ways of importing.

I still think that a radio-button below the file-list would be the best solution.
If "in-place" is active, the not-needed fields could be even hidden (not only greyed out):
![image](https://user-images.githubusercontent.com/5677563/110179411-51ac0c00-7e08-11eb-95b9-d4b553f4e095.png)

An alternative would be to use two buttons for the two usecases:
![image](https://user-images.githubusercontent.com/5677563/110179662-c0896500-7e08-11eb-991e-2e7cf2603a0a.png)

phweyland commented:
@quovadit. Thanks for your illustrated comments.
> "Keep this window open" is also great, but I would put that checkbox near the import button.

Seems to be a good idea. I'll  make a try.

> The only thing I still don't like is the choice between "in-place" and "copy".

I'm sorry for that but we use the standard preferences resources. That's a toggle checkout box or an enum box.
A radio button implementation as an alternative to the enum box should be possible. But not on my todo list, at least at the moment. BTW the horizontal layout is not the most common one and will probably add some complexity.

> An alternative would be to use two buttons for the two usecases:

I feel this dangerous. It's not a setting. And we cannot control the related sensitivity of other settings.

phweyland commented:
Based on your suggestions (except the radio buttons):

![image](https://user-images.githubusercontent.com/23012047/110189766-02ada900-7dff-11eb-87a8-7c2a853d3ebc.png)

- I'm not totally convinced by the 2 prefs on the same line. Saves a bit of height but does look very elegant to me. Let's see what the others think.
- the keep the window open button does it nicely.
- mode included in the import button. Probably more efficient but diving too much into the gtkdialog structure may not be a good practice.

quovadit commented:
> two buttons: I feel this dangerous. It's not a setting. And we cannot control the related sensitivity of other settings.

I don't see the problem if it's not a setting. What does 'related sensitivity of other settings' mean?

phweyland commented:
> What does 'related sensitivity of other settings' mean?

Sorry for the lack of clarity. In the following example the useless fields for in-situ import are disabled. 

![image](https://user-images.githubusercontent.com/23012047/110205546-521fc380-7e57-11eb-90d8-d5db7ba38cec.png)

quovadit commented:
I still don't see why two buttons are a problem. I did a quick hack and it works fine.

If opened via 'add in-place' button, the additional fields are not shown at all.
If opened via 'copy images' button, the additional fields are shown underneath the file-list.

In both cases the tab 'modes' is not needed anymore.

phweyland commented:
> In both cases the tab 'modes' is not needed anymore.

There is a misunderstanding here. For sure the user may need to tweak the patterns.

quovadit commented:
yes, the user may tweak the patterns, but only in copy-mode, and this should be done in the main-tab:

![image](https://user-images.githubusercontent.com/5677563/110207558-8eb5e400-7e84-11eb-8004-ed2ca9a38c1d.png)

phweyland commented:
Got it. That makes sense.
I would prefer to add the tab "patterns" in the case import-copy than losing space for the image list.
The key information I wanted to keep visible was the in-place or copy. But your suggestion offers a good solution.

quovadit commented:
I think at least the 'import job' should be visible on the main tab, because this will be different on every import.
So it's just one line we are losing.

AlicVB commented:
wahou, nice enhancements again !
some small comments (as usual) : 
- maybe it's just me, but I find the checkbox at the end surprising (but that's not a big problem either)
- not sure about the 2 buttons. I think that it would be better to let the choice in the import dialog so it can be changed on the fly (as it now). If I understand well last @quovadit proposal, we may just move the combobox for import methods (copy/in-place) to the main tab (and let everything else like it is)
That said, what I've said doesn't prevent to have 2 buttons in the lib panel that would just set different defaults values... but we need to have better buttons labels then ;)

johnny-bit commented:
> maybe it's just me, but I find the checkbox at the end surprising (but that's not a big problem either)

"keep the window open"? that's pretty much standard thing in some multi action dialog boxes imo.

quovadit commented:
I don't see the advantage of changing the choice on the fly.
But I think it's very important to see the two alternatives 'Add' or 'Copy' next to each other.
As this is not possible with the combo-box I strongly recommend the two buttons.
Having both (two buttons AND combo-box) feels quite strange for me. 

Some suggestions for the 2 button labels:
- add / copy
- in-place / copy
- add images / copy images
- add in-place / copy images
- import (in-place) / import (copy)

phweyland commented:
> I think at least the 'import job' should be visible on the main tab, because this will be different on every import.

Yes, but this can be the same for date/time override and some metadata as well.
Not sure there is a perfect solution. Everything on the first window consumes the list space. But having tabs hides the information and adds clicks.

Nilvus commented:
Thanks @phweyland for all your work here. Last update is much much better and it's good to see "keep the window open?" button. @AlicVB: It's better to see it where it is (at the bottom) as it's just near related buttons (cancel/import).

I like @quovadit last proposal. I agree with him that it's better to select at first what we want to do: add in-place or copy.

Near 99% of the times, I import from SD card, so copy mode. Clicking directly in good button will save just 3 clicks: when in import window, I actually need to click in modes tab (and I have to know that before), then change mode and after that going back to select images.

I'm also quite sure that this way we will have users "add in-place" images from SD card without seeing which mode is selected (even if it's written in import button), so quite risky. I already see complaints coming about skull heads instead of images after unplugging or removing images in their SD card.

Nilvus commented:
> > I think at least the 'import job' should be visible on the main tab, because this will be different on every import.
> 
> Yes, but this can be the same for date/time override and some metadata as well.
> Not sure there is a perfect solution. Everything on the first window consumes the list space. But having tabs hides the information and adds clicks.

Another thought: we could also use same expander in that dialog window than in import module in actual master. Just a thought!

quovadit commented:
> we could also use same expander in that dialog window

expander is better than tab, but still I think that 'import job' is the only setting you will use everytime (otherwise all filmrolls are called 'new')

date/time override is only necessary for special cases, and even now it is in the 'settings'-tab

AlicVB commented:
@Nilvus : I was speaking about the checkbox at the end of the label. no pb with the position inside the window.
For the 2 buttons, I must admit that you have good arguments here :)

And, as said before, +1 for the expander : it has the big advantage to be flexible enough for all needs :)

@quovadit : what about having a more meaningful default job name, which include the date for example ?

@phweyland : maybe not for this PR which is big enough, but I would say that a good enhancement for the "copy" method would be to show "on the fly" destination of each selected image altogether with a warning in case of name collisions (whatever the override setting)
I also have on my todo list a new import setting to compute thumbnails at import time (just say that so it can be eventually anticipated in the ui)

quovadit commented:
> what about having a more meaningful default job name, which include the date for example ?

the naming pattern by default is $(YEAR)-$(MONTH)-$(DAY)_$(JOBCODE) so it already includes the date, but still jobcode is unique by definition

phweyland commented:
> to show "on the fly" destination of each selected image altogether with a warning in case of name collisions (whatever the override setting)

The principle is nice.
But in the general case this implies to read the file to extract exif info. Applied to all images will probably slow down the process (not very different than reading the thumbnail).

Nilvus commented:
> @Nilvus : I was speaking about the checkbox at the end of the label. no pb with the position inside the window.
> For the 2 buttons, I must admit that you have good arguments here :)

Oh, sorry, didn't understand that. Indeed, I agree. @phweyland, could you put the checkbox in the left of "keep this window open" instead of the right? Would be better UI.

phweyland commented:
last changes:
- 2 buttons in import module
- 2 import dialogs
- checkbox / label swapped
- import-copy dialog default height increased to afford pattern settings

![image](https://user-images.githubusercontent.com/23012047/110243252-5839a100-7f38-11eb-8448-cd0afd598a26.png)

![image](https://user-images.githubusercontent.com/23012047/110243342-b070a300-7f38-11eb-8512-b7cb5ffcccb0.png)

TurboGit commented:
Then in the `import` module we may want to remove the **import** text in all buttons.

`[ in-place... ]` `[ copy...]`
`[ from camera ]`

quovadit commented:
I think a button should contain an action, so if we remove "import" (which is fine for me), than we should probably say 
- add in-place
- copy from disk
- copy from camera

TurboGit commented:
@quovadit : Agreed, a button should have an action. So I'll keep current names which are clear.

Nilvus commented:
@phweyland: you have succeed to disabled "file naming pattern" in dialog window if "keep original filename" is checked. It would be good to do same thing in import tab of general preferences. As you can see below, if same checkbox is checked, "file naming pattern" in preferences is not disabled:

![2021-03-07_19-49](https://user-images.githubusercontent.com/45535283/110250921-e7b17500-7f75-11eb-954f-e204076bfd0f.png)

phweyland commented:
Having moved metadata settings from import module to metadata dialog tab deprives the "import from camera" of any metadata control, and probably tethering as well (I don't know tethering and I haven't a compatible camera). 

As metadata are common to all import feature I think it would be better, at least at this stage, to let them in the import module.
Opinion ?

TurboGit commented:
> Having moved metadata settings from import module to metadata dialog

Sorry maybe there is some confusion on my side, I see no metadata settings on the metadata dialog.

Nilvus commented:
> Having moved metadata settings from import module to metadata dialog tab deprives the "import from camera" of any metadata control, and probably tethering as well (I don't know tethering and I haven't a compatible camera).
> 
> As metadata are common to all import feature I think it would be better, at least at this stage, to let them in the import module.
> Opinion ?

I would vote to let them in import module too. One place for something used for all import options is simpler.

phweyland commented:
> > Having moved metadata settings from import module to metadata dialog
> 
> Sorry maybe there is some confusion on my side, I see no metadata settings on the metadata dialog.

in import-copy and import-in-place dialogs there is a metadata tab (before metadata settings were in import module). Then these settings are missing for sure for import from camera and tethering. An alignment is always possible but in any case will take time and effort. And on my side I cannot test any import from a camera.

quovadit commented:
It's true, that the parameters are now missing for 'import from camera' and 'tethered shoot'.

Even now (3.4.1) some fields occur twice (preferences + import dialog), so I think this would be a good split:
- import module preferences (in hamburger menu ≡)
  - session options
    - base directory naming
    - sub directory naming
    - keep original filename
    - file naming pattern
  - initial import rating
  - defaults (can be temporarily overridden in import dialog)
    - creator, publisher, rights, notes
    - tags
- import in-place
  - main tab
    - import folders recursively
    - ignore JPEG files
  - details tab
    - title
    - description
    - creator, publisher, rights, notes
    - tags
- import copy
  - main tab
    - jobcode
  - details tab
    - override today's date
  - plus the same as 'in-place'
- import from camera
  - same as 'import copy' except 'import folders recursively'

phweyland commented:
- metadata back in import module
- all parameters captured in presets
- reset import module removed (not all presets are visible).

phweyland commented:
Thanks for all your kind suggestions.
I think I'm done feature wise.

quovadit commented:
I love the new feature, the wording of the 2 buttons is perfect now!

some remarks:
- after copy and import no image is shown because the wrong filmroll path is tried to open (source instead of destination?).
Clicking on the top-entry of the collect-images module opens the correct filmroll path and shows the imported images
- ignore jpg does not filter out the jpg files in the list anymore (they still are ignored on import). 
- override today's date is stored which makes no sense for me, because this changes every time of course
- in 3.4.1 it's possible to apply metadata + tags in the import dialog, so I think it's a step backwards to have this in parameters now
- why are the directory + file naming patterns still in the import dialog? (those are definitely one-time decisions)
- why is there still the import section in global 'darktable preferences' AND 'parameters' in the import module?
I would combine them in the module preferences (hamburger menu ≡)
- if you have the text on the right side of the checkbox, it's common to have the text clickable too

phweyland commented:
>     * after copy and import no image is shown because the wrong filmroll path is tried to open (source instead of destination?).

You are right, I've missed this case.

> ignore jpg does not filter out the jpg files in the list anymore

I cannot reproduce this. This work for me.

> override today's date is stored which makes no sense for me

Agreed, but that doesn't hurt neither.

> in 3.4.1 it's possible to apply metadata + tags in the import dialog, so I think it's a step backwards to have this in parameters now

This has been discussed already a lot. Back to the previous consensus.

> why are the directory + file naming patterns still in the import dialog?

Not sure to understand properly. But the copy and import dialog layout is your own proposal.

> why is there still the import section in global 'darktable preferences' AND 'parameters' in the import module?

?? there is no import section in global 'darktable preferences' anymore ....

> if you have the text on the right side of the checkbox, it's common to have the text clickable too

As all preferences, a double click resets the preference to its default value. If you want to add this feature globally, up to you to make a proposal. BTW I don't see why the text position would have an incidence here ...

Thanks for testing. Every bug detection is welcome.
But I would appreciate to stop the feature changes now.

phweyland commented:
> > ```
> > * after copy and import no image is shown because the wrong filmroll path is tried to open (source instead of destination?).
> > ```
> 
> 
> You are right, I've missed this case.

I've looked at this. All seems ok to me.
The right filmid is signaled as for the other import features.
On the other hand, as the importation is asynchronous, the user can switch to other activities and it would be not welcome to force the switch the selected collection to the newly imported. Additionally, "copy and import" can import several filmids...

phweyland commented:
last push: remove useless #include

Nilvus commented:
> > why is there still the import section in global 'darktable preferences' AND 'parameters' in the import module?
> 
> ?? there is no import section in global 'darktable preferences' anymore ....

@phweyland: I still have that import section in global darktable preferences in that PR (last commit):

![2021-03-08_21-23](https://user-images.githubusercontent.com/45535283/110377517-67117800-804c-11eb-84f5-f1106306d012.png)

And I always remove build folder and install folder before any new compile. And indeed, that tab is no more needed with that PR. So to be removed?

phweyland commented:
> > > why is there still the import section in global 'darktable preferences' AND 'parameters' in the import module?
> > 
> > 
> > ?? there is no import section in global 'darktable preferences' anymore ....
> 
> @phweyland: I still have that import section in global darktable preferences in that PR (last commit):

Right, the session options. But we have already removed the second paragraph of that section which was the import options (or something like that).

Nilvus commented:
> Right, the session options. But we have already removed the second paragraph of that section which was the import options (or something like that).

So do we keep that tab as you added those fields in import section? If we keep them, just to remember, if "keep original filename" is checked, it would be good to disable line pattern just below, like you did in same thing in that PR in dialog import window.

quovadit commented:
>> ignore jpg does not filter out the jpg files in the list anymore

> I cannot reproduce this.

In an earlier version the list was updated on-the-fly. But now although ignore jpg is selected, the jpgs are shown. 
![image](https://user-images.githubusercontent.com/5677563/110386590-c5e0ec80-8060-11eb-999b-223b81ba5930.png)


>> override today's date is stored which makes no sense for me

> Agreed,` but that doesn't hurt neither.

I think it does hurt, because it leads to wrong foldernames if I forget to remove the date from last import


>> why are the directory + file naming patterns still in the import dialog?

> the copy and import dialog layout is your own proposal.

Yes, It was my proposal to have all necessary information in the import dialog. But now we are forced to have some global settings because of 'import from camera' and 'tethering'.
That's why I would define one-time-decisions (like file naming patterns) in preferences, and information changing on every import (like title+description) in the import dialog. 


> So do we keep that tab as you added those fields in import section?

I would move all import settings including the session options from global preferences to the preferences (!) of the import module.

>> it's common to have the text clickable too

> But I would appreciate to stop the feature changes now.

Fine for me, it's no big deal anyway

phweyland commented:
> In an earlier version the list was updated on-the-fly. But now although ignore jpg is selected, the jpgs are shown.

Ok, upper case is not considered. bug, then.


> I think it does hurt, because it leads to wrong foldernames if I forget to remove the date from last import

Good point. Agreed.

> I would move all import settings including the session options from global preferences to the preferences (!) of the import module.

I have already proposed and coded all prefs in the presets menu of import module. But this has not been accepted by several people and I would not go backward now. I prefer as it is, even not perfect.
At least, where they are now you can have an eye on them. 

I'll fix the 2 first points. Thanks.

phweyland commented:
> So do we keep that tab as you added those fields in import section?

We cannot remove then form global preferences without jeopardize import form camera and tethering.

>  if "keep original filename" is checked, it would be good to disable line pattern just below, like you did in same thing in that PR in dialog import window.

I agree that would be fine, no doubt. But, sorry, this is not at all in that part of code.

phweyland commented:
- bugs above fixed.
- as I've just seen a complaint about dialog window height, I've set up an expander for patterns.

phweyland commented:
small change in ui preferences to control where the widget are added in the host grid.

@TurboGit , if no remaining bugs from others, I think this is good for your review.

TurboGit commented:
@phweyland : Ok, on my TODO list.

TurboGit commented:
I did a little test this morning and got a crash quickly. Two points:

- `[import in-place]`, I selected a directory with 63 pictures (not a lot) and clicked on the little eye icon on the top bar to display the thumbs, it took quite some time (~ 10s) without any display feedback. I would display all the thumbs one by one instead of waiting for all thumbs to be processed. I though dt was frozzen and I have a fast proc with OpenCL and a SSD hard drive. So with more pictures it will get even worse.

- I have then displayed/hidden some thumbs and finally clicked on `[import in-place]` in the dialog and  got a crash (reproducible). I attach the backtrace here:

```
#0  0x00007ffff76d83d0 in g_str_hash ()
    at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#1  0x00007ffff76d6d6b in g_hash_table_lookup ()
    at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007ffff7cc0470 in dt_confgen_get
    (name=<optimized out>, kind=kind@entry=DT_DEFAULT)
    at /home/obry/dev/builds/darktable/build/src/src/control/conf.c:574
#3  0x00007ffff7e0d139 in dt_gui_preferences_string_reset (widget=0x0)
    at /home/obry/dev/builds/darktable/build/src/src/gui/preferences.c:2489
#4  0x00007fff796e98b1 in _import_from_dialog_run
    (self=<optimized out>, self=<optimized out>)
    at /home/obry/dev/builds/darktable/build/src/src/libs/import.c:1023
#5  _lib_import_from_callback (widget=<optimized out>, self=<optimized out>)
    at /home/obry/dev/builds/darktable/build/src/src/libs/import.c:1072
#6  0x00007ffff6a0b2ee in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#7  0x00007ffff6a23a48 in g_signal_emit_valist ()
    at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#8  0x00007ffff6a23c3f in g_signal_emit ()
    at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#9  0x00007ffff701c310 in  () at /lib/x86_64-linux-gnu/libgtk-3.so.0
```

Restarting dt I see that some images where imported., so the crash is during import. Tested two times, first time 14 images imported and second time only 3.

I'm wondering if this new dialog should not be used by the `[import from camera]` too. But probably part of another PR. Let me know what you think about this.

- `[copy and import...]` tested and works, no crash. but at the end of the import I would expect the new collection to be selected which is not the case.

phweyland commented:
Thanks for testing.

>     * I selected a directory with 63 pictures (not a lot) and clicked on the little eye icon on the top bar to display the thumbs, it took quite some time (~ 10s) without any display feedback. 

Strange, I don't see that.

> I would display all the thumbs one by one instead of waiting for all thumbs to be processed.

It's what I've tried to code and that's what I see on my side... I'll check that.

> I attach the backtrace here:

Ugly ... this doesn't correspond to anything in the code. Maybe a memory corruption... 
I haven't tried to rebase with the current master. Did you rebase ?

> I'm wondering if this new dialog should not be used by the [import from camera] too. But probably part of another PR. Let me know what you think about this.

Import from camera and tethering are gphoto based. Seen the @jenshannoschwalm's comments that may not be too simple. But gphoto should be able to provide the thumbnails in a progressive way as we do here (what doesn't work for you :( )

> but at the end of the import I would expect the new collection to be selected which is not the case.

@quovadit has made this comment too. I've answered that I didn't think it was fine to force a collection change at the end of the import, as the user can have started another task in between. But it seems that the current import folder has a more subtle behavior. I'll check that.

phweyland commented:
> > ```
> > * I selected a directory with 63 pictures (not a lot) and clicked on the little eye icon on the top bar to display the thumbs, it took quite some time (~ 10s) without any display feedback. 
> > ```
> 
> 
> Strange, I don't see that

phweyland commented:
>  but at the end of the import I would expect the new collection to be selected which is not the case.

`import in-place` knows the future images folder (as current [folder...] import) and can update the collection accordingly before the import has finished (synchronously with the import button).
`copy and import` doesn't know at this time where the images will be.
Then do we want really to update the collection, asynchronously, potentially several minutes after starting the `copy and import`?

phweyland commented:
Rebased + fix crash on `import in-place`

I haven't found anything about the progressive thumbnail display which doesn't work for you.
Currently the thumbnail display loop is based on a 100 milliseconds time out. On my screen I'm getting a thumbnail every half a second or so ...

phweyland commented:
Lowering the timeout source priority makes the progressive thumb display work for me all the time. Hoping that solves the issue for you too.

TurboGit commented:
@phweyland : I'm a bit late to answer your questions. I'll test the new version.

And yes I'm always merging on top of master during my integration test. That's the only way I can ensure that no issue will be present in master, the whole point of my work :)

TurboGit commented:
For the display of newly imported collection I think we should not change this behavior. So what I would propose is to add a check box in the dialog (set ON by default) to jump to new collection or not.

phweyland commented:
> For the display of newly imported collection I think we should not change this behavior. So what I would propose is to add a check box in the dialog (set ON by default) to jump to new collection or not.

I'm not convinced. `Import from camera` doesn't do it neither, if I'm not mistaken. Synchronous and Asynchronous works should be treated accordingly.
I'll do as you suggest, at least to check how it does. In any case it won't be difficult to remove this if necessary.

quovadit commented:
The date/time column shows the modification-date/time of the file, and not the exif-date/time of the photo.
As this is also the default sort-order it leads to unexpected sequence-numbers.
I know there might not be exif-data (scanned images), but I think the main usecase will be to import photos with exif-data.

With modified date/time I still can't create a chronological sequence based on exif date/time, so the biggest advantage compared to the old "import folder" is gone...

Would it be possible to have 2 date/time columns, or switch between the 2 types?
Or choose automatically: default: exif date/time, no exif available -> modified date/time

In any case the column-header should show which date/time is used ("Exif date/time" or "Modified date/time")

phweyland commented:
> The date/time column shows the modification-date/time of the file, and not the exif-date/time of the photo.

Correct, that was described at the beginning.

> I think the main usecase will be to import photos with exif-data.

Agreed. And in this case the file date/time usually matches the exif one, right ?

> Would it be possible to have 2 date/time columns, or switch between the 2 types?
> Or choose automatically: default: exif date/time, no exif available -> modified date/time

This has already been discussed, hasn't it ? This is obviously possible to display the exif date/time instead or beside the file date/time. But this will me more or less as slow as displaying thumbnails.
So as for thumbnails the default should stay without.
But I agree this would be a useful option, which can be added later on.

phweyland commented:
I've jumped to new collection just after importing the first image.
This should be immediate and would not need any new preference.

phweyland commented:
Added tooltips on date/time (specifying 'file' date/time) and thumb column headers.

quovadit commented:
sorry I missed the discussion about exif before.
Fine for me to start that way, thanks for the tooltip, maybe you could also show the sorting arrow by default:
![image](https://user-images.githubusercontent.com/5677563/110842854-ebac0280-82a7-11eb-947a-acdf42b01174.png)

I love the collapsed parameters. But it's confusing that there are 'session parameters' in the import module module AND in the import dialog (named exactly the same but with different parameters).

phweyland commented:
I'm not at the computer at the moment, so I cannot check. If the 2 expanders have the same name that's my mistake. Should be different.

BTW, in the dialog window I've kept the word "session" to follow the global preferences, but I would find "patterns" more accurate.

phweyland commented:
Done:
- import module expander => parameters
- copy and import dialog expander => naming rules

Sorting:
More complex. The file enumeration doesn't guaranty the file order. To control this one more layer is required.
For the time being the order is manually controlled by the user, only...

phweyland commented:
> Sorting: More complex.

My bad. Was very simple in fact.

TurboGit commented on src/libs/import.c:
```diff
@@ -480,166 +508,581 @@ static void _lib_import_update_preview(GtkFileChooser *file_chooser, gpointer da
+    d->from.event = 0;
+    GtkTreeModel *model = GTK_TREE_MODEL(d->from.store);
+    GtkTreeIter iter;
+    gboolean valid = gtk_tree_model_get_iter_first(model, &iter);
```
given recent work from @ralfbrown  to replace while -> for I would rewrite this:

```
for(gboolean valid = gtk_tree_model_get_iter_first(model, &iter); valid; gtk_tree_model_iter_next(model, &iter))
{
   _thumb_set_in_listview(model, &iter, FALSE, self);
}
```

ralfbrown replied:
That would need to be
`for(gboolean valid = gtk_tree_model_get_iter_first(model, &iter); valid; valid = gtk_tree_model_iter_next(model, &iter))`
or it will loop forever.

TurboGit replied:
Yes of course! Thanks for catching that.

phweyland replied:
Done

TurboGit reviewed (commented):
Works for me now. No crash. Just a minor comment about a loop and we are good to go. A fantastic work again @phweyland. Thanks!

quovadit commented:
I have a very strange behaviour: when importing with the 'import from camera' button, all imported images are identical!

For example: I import 3 images 1812,1813,1814 - then I get 3x the image 1812 with identical exif information but different filenames:

When using a file naming pattern the filenames are 0001,0002,0003
If 'keep original filename' is checked, one image has the original filename 1812, the other images are named 0001,0002

![image](https://user-images.githubusercontent.com/5677563/110934194-fdd08400-832d-11eb-97ce-9ce4e85486b3.png)

![image](https://user-images.githubusercontent.com/5677563/110934198-0032de00-832e-11eb-8d44-6d8eb488922b.png)


It works correct with 3.4.1

phweyland commented:
> I have a very strange behaviour: when importing with the 'import from camera' button, all imported images are identical!

Let's check that...

phweyland commented:
> I have a very strange behaviour: when importing with the 'import from camera' button, all imported images are identical!

Good finding ! Fixed. (for once I'm not guilty :))

quovadit commented:
> Fixed.

works now, thanks

TurboGit reviewed (commented):
Looks good, I'll do another test soon.

quovadit commented:
what about changing the column-header from "file" + "date/time" --> "name" + "modified"
(as in the old 'import image' dialog)
![image](https://user-images.githubusercontent.com/5677563/111026102-d4286300-83e8-11eb-907c-a13380c85921.png)

![image](https://user-images.githubusercontent.com/5677563/111026157-32eddc80-83e9-11eb-8ae8-59a3945b051a.png)

phweyland commented:
> what about changing the column-header from "file" + "date/time" --> "name" + "modified"
> (as in the old 'import image' dialog)

Agreed, would be better.

quovadit commented:
> "I would like to see exact same thing for **import from camera**/card (mass media storage) button. Could you add that in that PR or in another?" (Nilvus)

This would be the last missing part of this outstanding new feature

phweyland commented:
> > "I would like to see exact same thing for **import from camera**/card (mass media storage) button. Could you add that in that PR or in another?" (Nilvus)
> 
> This would be the last missing part of this outstanding new feature

I understand the point, but the code is completely elsewhere and relies on gphoto. This is an other story.

TurboGit reviewed (approved):
All good to me now. Working fine and a good revamp of the import dialog. I'll merge now to get more field testing.

One point maybe to tackle in another PR. Do you think it would be possible for the import in-place to list only pictures not already imported ?

AlicVB commented:
hmm... reimporting picture can be needed in some cases. Maybe mark them differently and don't select them by default ? Ideally we can have another column so it can be sortable too (but maybe I ask too much :)

TurboGit commented:
@AlicVB : I don't understand, reimporting is a nop, no?

AlicVB commented:
no, iirc reimporting reload xmp files. (I vaguely remember to have used that with my xmp backups after having screwed everything with a bad hacking attempt long time ago)
to be checked, my memory do funny things sometimes :)

phweyland commented:
> Do you think it would be possible for the import in-place to list only pictures not already imported ?

Possible, yes. As we `import in-place`, it's easy to check if the film_roll already exists and eventually check the files. A checkbox like "ignore jpegs" should work.

But I already see the next request: could we do the same for `copy and import` ?
This would be more difficult as we have to read the exif information. And then that could detect a collision but not an already imported image... In that case, it would be probably better to to show another column with the calculated filenames. Up to the user to adjust the list and/or the rules.

Another potential upcoming feature: add more naming rules. Potentially:
- overwrite (currently the sequence number is incremented as long as there is a conflict). 
- sequence start number

> reimporting is a nop, no?

Unless there are new xmp... (I just see you have already answered yourself :))

phweyland commented:
I'm looking at `import from camera` and try to understand the code. So far I don't see any roadblock.
The only issue I've found is about the file datetime. It seems that gphoto API doesn't provide it. @jenshannoschwalm. Your experience may save a lot of time. If you have any advice it is welcome. TIA.

jenshannoschwalm commented:
How can i help here? Anything specific?

I didn't follow the complete conversation ... Some first ideas

1. Yes, there seems to be no gphoto specific support for for file datetime
2. The major issue i found with gphoto is mem management (therefore the trick with keeping data for previews)
3. It would be very good to disable import via `import_from_camera` for any readable usb device as that is supported in a better way right now :-)

 Or would you like me to take care of (3)?

phweyland commented:
Thanks @jenshannoschwalm 

2. I've seen that, as soon as a preview has been requested, the corresponding data are kept until the camera is destroyed. I'll keep this method. Comes up the idea to reuse these data (for example the user request the thumbnail for an image, hides it and requests it again...). A possible improvement when the basics work ...
3. Good idea ! If I understand properly this corresponds to the "mass storage" "import from camera", right ?

jenshannoschwalm commented:
> if you want all thumbnails and pre-rendered images in best quality you should choose the always option\n(more comments in the manual)

exactly. i guess you have already noticed the work-around :-)

probably it's relatively easy to fix via rejecting in the background thread first ...
